workshop.md

```diff
@@ -91,163 +95,265 @@ Once you are in the new Lakehouse, create a new notebook by selecting `Open Note
-url = f"{LEARN_GITHUB_BASE}/wwl/get-started-kusto-fabric/index.yml"
+#### 2. Define Helper Functions
 
-# load module metadata
-response = requests.get(url)
-index_data = yaml.safe_load(response.content)
+Now we'll define all the helper functions we'll use throughout the tutorial.
 
-# print the module metadata
-index_data
+```python
+# Helper Functions
+
+def parse_github_url(url):
+    """
+    Parses a GitHub URL and extracts the owner, repository name, branch, and path.
+    """
+    pattern = r"https://github\.com/(?P<owner>[^/]+)/(?P<repo>[^/]+)/tree/(?P<branch>[^/]+)/(?P<path>.+)"
+    match = re.match(pattern, url)
+    if match:
+        return match.groupdict()
+    else:
+        raise ValueError("Invalid GitHub URL format.")
+
+def download_and_save_file(file_url, lakehouse_save_path):
+    """
+    Downloads a file from GitHub and saves it to the specified path in the Lakehouse.
+    """
+    try:
+        response = requests.get(file_url)
+        if response.status_code == 200:
+            # Create directories in Lakehouse if they don't exist
+            dir_path = os.path.dirname(lakehouse_save_path)
+            mssparkutils.fs.mkdirs(dir_path)
+            # Save the file content to Lakehouse
+            mssparkutils.fs.put(lakehouse_save_path, response.text, overwrite=True)
+            print(f"Successfully downloaded {lakehouse_save_path}")
+        else:
+            print(f"Failed to download {file_url}: HTTP {response.status_code}")
+    except Exception as e:
+        print(f"Error downloading {file_url}: {e}")
+
+def download_contents(contents, base_save_path):
+    """
+    Recursively downloads contents from a GitHub repository and saves them to the Lakehouse.
+    """
+    for item in contents:
+        if item['type'] == 'file':
+            file_name = item['name']
+            download_url = item['download_url']
+            # Compute the relative path of the file within the module
+            relative_path = item['path'].replace(REPO_PATH + '/', '')
+            # Construct the Lakehouse save path
+            lakehouse_save_path = f"{base_save_path}/{relative_path}"
+            # Download and save the file
+            download_and_save_file(download_url, lakehouse_save_path)
+        elif item['type'] == 'dir':
+            # Handle subdirectories recursively
+            sub_dir_path = item['path']
+            sub_api_url = item['url']
+            sub_response = requests.get(sub_api_url)
+            if sub_response.status_code == 200:
+                sub_contents = sub_response.json()
+                # Compute the new base save path for the subdirectory
+                sub_base_save_path = f"{base_save_path}/{item['name']}"
+                download_contents(sub_contents, sub_base_save_path)
+            else:
+                print(f"Failed to list contents of {sub_dir_path}: HTTP {sub_response.status_code}")
 ```
 
-Run the cell.
+**Explanation of Helper Functions:**
```

Comment:
In general the code above should be split into what the user will paste in each notebook cell. Think of a tutorial. So there should be a paragraph of explanation either before the sample code, or after, and then the code. Then the other function explanation, the code, and so on. Otherwise people will either be puzzled by the bunch of code, could skip it, jump back and forth between bunch of code/explanation, etc. Think about UX but for someone learning in this case.